File: src/main/java/algorithms/java/easy/RemoveElement.java
package algorithms.java.easy;

/**
 * #27. Remove Element
 * https://leetcode-cn.com/problems/remove-element/
 * 思路：通过变量i 重排数组，通过 numbers[j] == val 跳过指定元素
 * @author xiaobin
 */
public class RemoveElement {

    public static int removeElement(int[] numbers, int val) {
//        int i = 0;
//        int n = nums.length;
//        while (i < n) {
//            if(nums[i] == val) {
//                nums[i] = nums[n - 1];
//                n--;
//            } else {
//                i++;
//            }
//        }
//        return i;

        // 通过变量i 重排数组，通过 nums[j] == val 跳过指定元素，获取移除元素后的新数组
        int i = 0;
        for (int j = 0; j < numbers.length; j++) {
            if (numbers[j] == val) { continue; }
            numbers[i] = numbers[j];
            i++;
        }
        return i;
    }

    public static void main(String[] args) {
        int[] numbers = {0,1,2,2,3,0,4,2};
        int newLength = RemoveElement.removeElement(numbers, 2);
        System.out.println("newLength=" + newLength);
    }
}

File: ruby/remove_element.rb
=begin
    
LC#27 移除元素 Link：https://leetcode-cn.com/problems/remove-element/

思路：类似 LC#26
1：双指针遍历，j负责移动， i负责标记
2：当发现 nums[j] == val 时候，将元素位置移动到 [i]
3：
=end
def remove_element(nums, val)
    i = 0
    nums.each_index do |j| 
        if nums[j] != val
            nums[i] = nums[j]
            i+=1
        end
    end
    i
end

r = remove_element([3,2,2,3], 2)
p "rr> ", r